manifests/config.pp

```diff
@@ -1,34 +1,51 @@
 # Private class
-class haproxy::config inherits haproxy {
+define haproxy::config (
+  $instance_name,
+  $config_dir,
+  $config_file,
+  $global_options,
+  $defaults_options,
+  $custom_fragment = undef,  # Required for Puppet 2.7 compatibility. Remove default when deprecated.
+) {
   if $caller_module_name != $module_name {
     fail("Use of private class ${name} by ${caller_module_name}")
   }
 
-  concat { $haproxy::config_file:
+  if $config_dir != undef {
```

Comment:
This is a private define. The caller will always set config_dir if there is a parent directory that needs to be created.  Otherwise it is left undef and is not created.